File: labs/mobile/donor-archive/tunnel/StaticCidrPool.kt
// Copyright 2024-2026 LumiNet Authors. All Rights Reserved.
package com.luminet.android.tunnel

data class StaticClientRecord(
    val clientId: String,
    val allocatedIp: String,
    val publicKey: String,
    var isRevoked: Boolean = false
)

class StaticCidrPool(private val basePrefix: String = "10.66.0.") {
    private var nextHost = 2
    private val allocated = mutableMapOf<String, StaticClientRecord>()
    private val revokedKeys = mutableSetOf<String>()
    private val lock = Any()

    fun allocateClient(clientId: String, publicKey: String): StaticClientRecord? = synchronized(lock) {
        if (nextHost >= 254) return null
        val ip = "$basePrefix$nextHost"
        nextHost++

        val record = StaticClientRecord(clientId, ip, publicKey)
        allocated[ip] = record
        record
    }

    fun revokeClient(ip: String): Boolean = synchronized(lock) {
        val client = allocated[ip] ?: return false
        client.isRevoked = true
        revokedKeys.add(client.publicKey)
        true
    }

    fun isKeyRevoked(publicKey: String): Boolean = synchronized(lock) {
        revokedKeys.contains(publicKey)
    }
}

File: labs/control-ui/api-shadow/static_cidr_pool.ts
export interface StaticClientRecord {
  clientId: string;
  allocatedIp: string;
  publicKey: string;
  isRevoked: boolean;
}

export class StaticCidrPool {
  private nextHost = 2;
  private readonly allocated = new Map<string, StaticClientRecord>();
  private readonly revokedKeys = new Set<string>();
  private readonly basePrefix;
  constructor(basePrefix = '10.66.0.') {
    this.basePrefix = basePrefix;
  }

  public allocateClient(clientId: string, publicKey: string): StaticClientRecord | null {
    if (this.nextHost >= 254) return null;
    const ip = `${this.basePrefix}${this.nextHost}`;
    this.nextHost++;

    const record: StaticClientRecord = {
      clientId,
      allocatedIp: ip,
      publicKey,
      isRevoked: false,
    };
    this.allocated.set(ip, record);
    return record;
  }

  public revokeClient(ip: string): boolean {
    const client = this.allocated.get(ip);
    if (!client) return false;
    client.isRevoked = true;
    this.revokedKeys.add(client.publicKey);
    return true;
  }

  public isKeyRevoked(publicKey: string): boolean {
    return this.revokedKeys.has(publicKey);
  }
}
